packages/protocol/test/identity/escrow.ts

```diff
@@ -232,238 +362,410 @@ contract('Escrow', (accounts: string[]) => {
+          )
+        })
+
+        it('should allow users to withdraw after completing attestations', async () => {
+          await withdrawAndCheckState(
+            sender,
+            receiver,
+            aPhoneHash,
+            aValue,
+            uniquePaymentIDWithdraw,
+            minAttestations,
+            [],
+            []
+          )
+        })
+        it('should not allow a user to withdraw a payment if they have fewer than minAttestations', async () => {
+          await assertRevertWithReason(
+            withdrawAndCheckState(
+              sender,
+              receiver,
+              aPhoneHash,
+              aValue,
+              uniquePaymentIDWithdraw,
+              minAttestations - 1,
+              [],
+              []
+            ),
+            'This account does not have enough attestations to withdraw this payment.'
+          )
+        })
+        it("should withdraw properly when sender's second payment has an identifier", async () => {
+          await mintAndTransfer(
+            sender,
+            aPhoneHash,
+            aValue,
+            oneDayInSecs,
+            anotherWithdrawKeyAddress,
+            0
+          )
+          await withdrawAndCheckState(
+            sender,
+            receiver,
+            aPhoneHash,
+            aValue,
+            uniquePaymentIDWithdraw,
+            minAttestations,
+            [anotherWithdrawKeyAddress],
+            [anotherWithdrawKeyAddress]
+          )
+        })
       })
     })
 
     describe('#revoke()', () => {
       let uniquePaymentIDRevoke: string
       let parsedSig1: any
 
-      beforeEach(async () => {
-        await doubleTransfer()
-        uniquePaymentIDRevoke = withdrawKeyAddress
-        parsedSig1 = await getParsedSignatureOfAddress(web3, accounts[2], accounts[5])
-      })
-
-      it('should allow sender to redeem payment after payment has expired', async () => {
-        await timeTravel(oneDayInSecs, web3)
-        const escrowedPaymentBefore = await getEscrowedPayment(uniquePaymentIDRevoke, escrow)
-
-        await escrow.revoke(uniquePaymentIDRevoke, { from: sender })
-
-        const sentPaymentsAfterRevoke = await escrow.getSentPaymentIds(sender)
-        const receivedPaymentsAfterRevoke = await escrow.getReceivedPaymentIds(aPhoneHash)
-
-        assert.notInclude(
-          sentPaymentsAfterRevoke,
-          uniquePaymentIDRevoke,
-          "Should have deleted this escrowed payment from sender's sentPaymentIds list after revoke"
-        )
-
-        assert.notInclude(
-          receivedPaymentsAfterRevoke,
-          uniquePaymentIDRevoke,
-          "Should have deleted this escrowed payment from receiver's receivedPaymentIds list after revoke"
-        )
-
-        const escrowedPaymentAfter = await getEscrowedPayment(uniquePaymentIDRevoke, escrow)
+      const identifiers = [NULL_BYTES32, aPhoneHash]
+      identifiers.forEach((identifier) => {
+        describe(`when identifier is ${
```

Comment:
death to duplicate code haha 🚀

Reply:
code === duplicate ? 'not' : 'ok'

Reply:
🙈  troo
yeah even the transfer/withdraw function that I'm now using is not ideal test structure IMO and the next step of refactoring (post-audit deadline) would be simplifying + breaking apart from single maaaassivveee test cases -> smaller cases with fewer asserts hehe hopefully also more readable 🙈 😬

Reply:
I think this already pretty good tbh haha. Don't want to discourage you though! Never hurts to improve tests especially when it comes to smart contracts that hold user funds